File: example/custom.lua
-- 自定义全局变量
FirstSwitch = true; -- 起手标记
HighPing = false;   -- 高延迟标记, 高延迟下, 有光明相起手无法使用 0 容错打法.

function Init()
    -- 每次战斗开始前, 重置全局变量
    FirstSwitch = true;
    HighPing = false;
end

MacroNum = 4; -- 自定义宏数量.

-- 宏0, 用于起手和进入大齐光的准备.
function Macro0(player)
    if player.nCurrentMoonEnergy >= 10000 or player.nCurrentMoonEnergy <= 2000 then
        player:cast(3967); -- /cast [moon>99&sun<21] 净世破魔击
    end
    player:cast(3979);     -- /cast 驱夜断愁
    player:cast(3963);     -- /cast 烈日斩
    if player.nCurrentSunEnergy >= 10000 and player.nCurrentMoonEnergy == 8000 then
        -- 切换至 1 号宏
        player.macroIdx = 1; -- /switch [sun>99&moon=80] 1
        if FirstSwitch then
            -- 起手时, 早一些释放暗尘弥散, 以避免第二轮卡隐身 CD
            player.delayCustom = 300; -- 将下一次释放宏的时机设置至 300 tick 后. 1024 tick 为 1 秒.
            -- 如果不设置 player.delayCustom, 那么下一次释放宏的时机会是 公共 CD 冷却完成后 + 基础延迟 + 随机延迟
            FirstSwitch = false;      -- 将起手标记置为 false
        end
    end
end

-- 宏1, 用于 齐光3 前半部分.
function Macro1(player)
    player:cast(3974); -- /cast 暗尘弥散
    -- 如果隐身没好, 等隐身
    if not player:buffExist(25731, 1) and not player:buffExist(25721, 3) then
        -- 在本宏的全部过程中, 要么有 降灵尊 buff (25731,1), 要么有 齐光3 buff (25721,3). 如果两者都没有, 说明隐身没有成功释放.
        return; -- 直接返回, 不进行后续操作. 由于公共 CD 没有被触发, 因此下一次释放宏的时机为 基础延迟 + 随机延迟 后.
    end

    if player:buffExist(25721, 3) and not player:buffExist(25716, 0) and player.nCurrentMoonEnergy >= 10000 then
        player:cast(3969);        -- /cast [buff:齐光3 & nobuff:悬象(包括日和月) & moon>99] 光明相
    end
    player:cast(34347);           -- /cast 悬象著明
    if player:buffExist(25716, 0) then
        player:itemUse(1, 38789); -- /cast [buff:悬象(包括日和月)] 吹香雪(特效腰坠)
    end
    player:cast(3966);            -- /cast 生死劫
    -- 如果生死劫没好, 等生死劫
    if player:buffExist(25721, 1) then
        -- 生死劫会将齐光升至 2 阶. 如果此时仍有 齐光1 的 buff, 说明生死劫被卡 CD.
        return; -- 直接返回, 不进行后续操作. 由于公共 CD 没有被触发, 因此下一次释放宏的时机为 基础延迟 + 随机延迟 后.
    end

    if player.nCurrentMoonEnergy <= 4000 then
        player:cast(22890); -- /cast [moon<41] 诛邪镇魔
    end
    player:cast(3967);      -- /cast 净世破魔击
    if player.nCurrentMoonEnergy <= 4000 then
        player:cast(3979);  -- /cast [moon<41] 驱夜断愁
    end
    player:cast(3960);      -- /cast 银月斩
    player:cast(3963);      -- /cast 烈日斩

    if player:buffExist(25721, 3) and not player:buffExist(25716, 0) and player.nCurrentMoonEnergy == 0 then
        -- 切换至2号宏
        player.macroIdx = 2; -- /switch [buff:齐光3 & nobuff:悬象(包括日和月) & moon=0] 2
    end
end

-- 宏2, 用于 齐光3 后半部分.
function Macro2(player)
    if player.nCurrentSunEnergy >= 10000 and player.nCurrentMoonEnergy >= 10000 then
        -- sun>=100&moon>=100, 说明 齐光3 结束返灵了. 走 齐光3 结束的流程.
        if HighPing then
            -- 处理高延迟情况.
            -- 高延迟下, 即使 齐光3 结束返灵, 该日斩也必须打出, 否则会在小齐光部分开头卡月斩 CD.
            player:cast(3963); -- /cast 烈日斩
            if player.nCurrentSunEnergy < 14000 then
                -- 日斩打出 + 齐光3 结束返还 100 能量, nCurrentSunEnergy < 14000 说明日斩未成功释放.
                return; -- 直接返回, 不进行后续操作. 由于公共 CD 没有被触发, 因此下一次释放宏的时机为 基础延迟 + 随机延迟 后.
            end
        end
        -- 如果没有返回, 说明日斩成功释放, 可以继续走 齐光3 结束的流程.
        HighPing = false;    -- 将高延迟标记重置为 false
        player.macroIdx = 3; -- 切换至 3 号宏
        Macro3();            -- 执行一次 3 号宏
        return;              -- 直接返回, 不进行后续操作
        -- 在宏开始时进行判断的原因是, 导致切换条件 (sun>=100&moon>=100) 的事件 (齐光3 结束返灵) 是发生在公共 CD 中的,
        -- 而不是发生在宏内由宏内的技能导致的. 因此, 必须在宏开始时进行判断, 而不能在宏结束时进行判断.
    end
    if player:buffTimeLeftTick(25721, 3) < 5 * (1024 * 15 / 16 + player.delayBase + player.delayRand) and player:buffExist(9909, 0) then
        -- 齐光剩余时间小于 5 个 GCD, 且手上的诛邪还没打出去, 换高延迟打法
        HighPing = true; -- 将高延迟标记置为 true
    end

    -- 高延迟打法
    if HighPing then
        -- 当 highPing 被标记时, 一定已经到了最后 5 个技能. 所以这里可以省去判定条件, 直接按顺序释放技能.
        player:cast(3963);  -- /cast 烈日斩
        player:cast(22890); -- /cast 诛邪镇魔
        player:cast(3960);  -- /cast 银月斩
        player:cast(3967);  -- /cast 净世破魔击
    end

    -- 正常打法
    if player.nCurrentMoonEnergy <= 4000 then
        player:cast(22890); -- /cast [moon<41] 诛邪镇魔
    end
    player:cast(3967);      -- /cast 净世破魔击
    if player.nCurrentMoonEnergy <= 4000 then
        player:cast(3979);  -- /cast [moon<41] 驱夜断愁
    end
    player:cast(3963);      -- /cast 烈日斩
    player:cast(3960);      -- /cast 银月斩
end

-- 宏3, 用于 齐光3 结束后直至 齐光2 的结束阶段.
function Macro3(player)
    if player.nCurrentMoonEnergy >= 6000 then
        player:cast(22890); -- /cast [moon>59] 诛邪镇魔
    end
    if player.nCurrentSunEnergy >= 10000 then
        player:cast(3966); -- /cast [sun>99] 生死劫
    end
    player:cast(3967);     -- /cast 净世破魔击
    player:cast(3960);     -- /cast 银月斩
    if player.nCurrentMoonEnergy >= 6000 or player.nCurrentSunEnergy >= 6000 then
        player:cast(3963); -- /cast [moon>59&sun>59] 烈日斩
    end
    if player.nCurrentMoonEnergy == 6000 and player.nCurrentSunEnergy == 4000 then
        player:cast(3962); -- /cast [moon=60&sun=40] 赤日轮
    end
    if player.nCurrentMoonEnergy >= 10000 and player.nCurrentSunEnergy < 10000 and not player:buffExist(25721, 0) then
        -- 切换至 0 号宏
        player.macroIdx = 0; -- /switch [moon>99&sun<100&nobuff:齐光] 0
        -- 在结尾进行判断的原因是, 导致切换条件 (moon>99&sun<100&nobuff:齐光) 的事件 (银月斩的释放) 是发生在宏内的. 因此, 可以在宏结束时进行判断.
    end
end

File: source/src/frame/custom.cpp
#include "frame/character/player.h"
#include <format>
#include <string>
#include <unordered_map>
#include <vector>

using namespace jx3calc;
using namespace frame;

std::shared_ptr<CustomLua> CustomLua::get(const std::string &script) {
    if (!mapCustomLua.contains(script)) {
        mapCustomLua.emplace(script, std::make_shared<CustomLua>(script));
    }
    return mapCustomLua.at(script);
}

void CustomLua::cancel(const std::string &script) {
    if (mapCustomLua.contains(script)) {
        mapCustomLua.erase(script);
    }
}

static std::vector<std::string> split(const std::string &s, char delimiter) {
    std::vector<std::string> tokens;

    const char *p = s.c_str(), *q = s.c_str();
    while (*p != '\0') {
        if (*p == delimiter) {
            tokens.emplace_back(q, p - q);
            q = p + 1;
        }
        p++;
    }
    tokens.emplace_back(q, p - q);
    return tokens;
}

static std::string convertRValBuff(const char *start, size_t len) {
    if (0 == len) {
        return ">0";
    }
    return std::string(start, len);
}

static std::string convertRValSunMoon(const char *start, size_t len) {
    if (0 == len) {
        return ">0";
    }
    std::string str {start + 1, len - 1};
    if (str == "sun") {
        str = "player.nCurrentSunEnergy";
    } else if (str == "moon") {
        str = "player.nCurrentMoonEnergy";
    } else {
        try {
            str = std::to_string(std::stoi(str) * 100);
        } catch (...) {
        }
    }
    return std::format("{}{}", *start, str);
}

static std::string convertCondition(const char *cond) {
    // cond: buff:日月同辉=3&nobuff:诛邪镇魔|sun<20
    class CondType {
    public:
        const size_t      len;
        const std::string name;
        std::string (*convert)(const char *start, size_t len);
    };
    static const std::unordered_map<std::string, CondType> condMap = {
        {"buff", {4, "player:macroBuff", convertRValBuff}},
        {"nobuff", {6, "player:macroNoBuff", nullptr}},
        {"bufftime", {8, "player:macroBufftime", nullptr}},
        {"tbuff", {5, "player:macroTBuff", convertRValBuff}},
        {"tnobuff", {7, "player:macroTNoBuff", nullptr}},
        {"tbufftime", {9, "player:macroTBufftime", nullptr}},
        {"sun", {3, "player.nCurrentSunEnergy", convertRValSunMoon}},
        {"moon", {4, "player.nCurrentMoonEnergy", convertRValSunMoon}},
    };
    if (*cond == '\0') {
        return "";
    }
    // 检查条件类型
    const CondType *type = nullptr;
    for (const auto &[key, value] : condMap) {
        if (strncmp(cond, key.c_str(), key.length()) == 0) {
            type = &value;
            break;
        }
    }
    if (type == nullptr) {
        return "";
    }

    // 计算
    const char *end = cond;
    while (*end != '&' && *end != '|' && *end != '\0')
        end++; // end 指向 & 或 | 或 \0
    const char *argument = cond + type->len;
    const char *op       = argument;
    while (op < end && *op != '=' && *op != '<' && *op != '>')
        op++;
    return std::format(
        "({}{}{}{}{}{})",
        type->name,    // 函数调用或成员变量
        argument == op // 如果是带参数的函数调用, 补充参数
            ? ""
            : "(\"" + std::string(argument + 1, op - argument - 1) + "\")",
        *op == '=' ? "=" : "",   // 如果操作符是等号, 补一个等号
        type->convert == nullptr // 操作符和操作符右值
            ? std::string(op, end - op)
            : type->convert(op, end - op),
        *end == '&' ? " and " : (*end == '|' ? " or " : ""), // 与下一个条件的连接符
        convertCondition(end + 1)
    );
}

std::string CustomLua::parse(std::vector<std::string> macroList) {
    std::string result;
    result += std::format("function Init() end;\nMacroNum = {};\n", macroList.size());
    for (size_t i = 0; i < macroList.size(); i++) {
        result += std::format("function Macro{}(player)\n", i);
        auto lines = split(macroList[i], '\n'); // 将字符串以 \n 为分隔符进行分割
        for (const auto &line : lines) {
            auto words     = split(line, ' '); // 将字符串以空格为分隔符进行分割
            bool condition = words.size() == 3 && words[1].starts_with("[") && words[1].ends_with("]");
            if (words[0] == "/cast") {
                if (condition) {
                    result += "    if" + convertCondition(words[1].substr(1, words[1].size() - 2).c_str()) + " then\n";
                    result += "        player:macroSkillCast(\"" + words[2] + "\");\n";
                    result += "    end\n";
                } else {
                    result += "    player:macroSkillCast(\"" + words[1] + "\");\n";
                }
            } else if (words[0] == "/switch") {
                int idx = condition ? 2 : 1;
                try {
                    int newIdx = std::stoi(words[idx]);
                    if (newIdx < 0 || newIdx >= macroList.size()) {
                        throw std::exception();
                    }
                    if (condition) {
                        result += "    if" + convertCondition(words[1].substr(1, words[1].size() - 2).c_str()) + " then\n";
                        result += "        player.macroIdx = " + std::to_string(newIdx) + ";\n";
                        result += "    end\n";
                    } else {
                        result += "        player.macroIdx = " + std::to_string(newIdx) + ";\n";
                    }
                } catch (...) {
                }
            }
        }
        result += "end\n\n";
    }
    return result;
}

static void constructorBefore(CustomLua *self);
static void constructorAfter(CustomLua *self);

CustomLua::CustomLua(const std::string &script) {
    constructorBefore(this);
    lua.script(script);
    constructorAfter(this);
}

static void constructorBefore(CustomLua *self) {
    self->lua.open_libraries(sol::lib::base, sol::lib::math);

    auto chAttr = self->lua.new_usertype<ChAttr>("ChAttr");

    chAttr["atVitalityBase"]                   = &ChAttr::atVitalityBase;
    chAttr["atStrengthBase"]                   = &ChAttr::atStrengthBase;
    chAttr["atAgilityBase"]                    = &ChAttr::atAgilityBase;
    chAttr["atSpiritBase"]                     = &ChAttr::atSpiritBase;
    chAttr["atSpunkBase"]                      = &ChAttr::atSpunkBase;
    chAttr["atPhysicsAttackPowerBase"]         = &ChAttr::atPhysicsAttackPowerBase;
    chAttr["atSolarAttackPowerBase"]           = &ChAttr::atSolarAttackPowerBase;
    chAttr["atLunarAttackPowerBase"]           = &ChAttr::atLunarAttackPowerBase;
    chAttr["atNeutralAttackPowerBase"]         = &ChAttr::atNeutralAttackPowerBase;
    chAttr["atPoisonAttackPowerBase"]          = &ChAttr::atPoisonAttackPowerBase;
    chAttr["atTherapyPowerBase"]               = &ChAttr::atTherapyPowerBase;
    chAttr["atPhysicsCriticalStrike"]          = &ChAttr::atPhysicsCriticalStrike;
    chAttr["atSolarCriticalStrike"]            = &ChAttr::atSolarCriticalStrike;
    chAttr["atLunarCriticalStrike"]            = &ChAttr::atLunarCriticalStrike;
    chAttr["atNeutralCriticalStrike"]          = &ChAttr::atNeutralCriticalStrike;
    chAttr["atPoisonCriticalStrike"]           = &ChAttr::atPoisonCriticalStrike;
    chAttr["atPhysicsCriticalDamagePowerBase"] = &ChAttr::atPhysicsCriticalDamagePowerBase;
    chAttr["atSolarCriticalDamagePowerBase"]   = &ChAttr::atSolarCriticalDamagePowerBase;
    chAttr["atLunarCriticalDamagePowerBase"]   = &ChAttr::atLunarCriticalDamagePowerBase;
    chAttr["atNeutralCriticalDamagePowerBase"] = &ChAttr::atNeutralCriticalDamagePowerBase;
    chAttr["atPoisonCriticalDamagePowerBase"]  = &ChAttr::atPoisonCriticalDamagePowerBase;
    chAttr["atPhysicsOvercomeBase"]            = &ChAttr::atPhysicsOvercomeBase;
    chAttr["atSolarOvercomeBase"]              = &ChAttr::atSolarOvercomeBase;
    chAttr["atLunarOvercomeBase"]              = &ChAttr::atLunarOvercomeBase;
    chAttr["atNeutralOvercomeBase"]            = &ChAttr::atNeutralOvercomeBase;
    chAttr["atPoisonOvercomeBase"]             = &ChAttr::atPoisonOvercomeBase;
    chAttr["atSurplusValueBase"]               = &ChAttr::atSurplusValueBase;
    chAttr["atStrainBase"]                     = &ChAttr::atStrainBase;
    chAttr["atHasteBase"]                      = &ChAttr::atHasteBase;
    chAttr["atPhysicsShieldBase"]              = &ChAttr::atPhysicsShieldBase;
    chAttr["atSolarMagicShieldBase"]           = &ChAttr::atSolarMagicShieldBase;
    chAttr["atLunarMagicShieldBase"]           = &ChAttr::atLunarMagicShieldBase;
    chAttr["atNeutralMagicShieldBase"]         = &ChAttr::atNeutralMagicShieldBase;
    chAttr["atPoisonMagicShieldBase"]          = &ChAttr::atPoisonMagicShieldBase;
    chAttr["atDodge"]                          = &ChAttr::atDodge;
    chAttr["atParryBase"]                      = &ChAttr::atParryBase;
    chAttr["atParryValueBase"]                 = &ChAttr::atParryValueBase;
    chAttr["atToughnessBase"]                  = &ChAttr::atToughnessBase;
    chAttr["atDecriticalDamagePowerBase"]      = &ChAttr::atDecriticalDamagePowerBase;
    chAttr["atMeleeWeaponDamageBase"]          = &ChAttr::atMeleeWeaponDamageBase;
    chAttr["atMeleeWeaponDamageRand"]          = &ChAttr::atMeleeWeaponDamageRand;

    auto buff = self->lua.new_usertype<BuffItem>("Buff");

    buff["nLevel"]           = &BuffItem::nLevel;
    buff["nIndex"]           = &BuffItem::nIndex;
    buff["nStackNum"]        = &BuffItem::nStackNum;
    buff["nLeftActiveCount"] = &BuffItem::nLeftActiveCount;
    buff["nNextActiveFrame"] = &BuffItem::nNextActiveFrame;
    buff["nCustomValue"]     = &BuffItem::nCustomValue;

    auto player = self->lua.new_usertype<Player>("Player");

    player["chAttr"]                 = &Player::chAttr;
    player["characterGetEmployer"]   = &Player::characterGetEmployer;
    player["characterGetSelect"]     = &Player::characterGetSelect;
    player["characterGetTargetID"]   = &Player::characterGetTargetID;
    player["buffExist"]              = &Player::buffExist;
    player["buffGet"]                = &Player::buffGet;
    player["buffGetByOwner"]         = &Player::buffGetByOwner;
    player["buffAdd"]                = &Player::buffAddOptional;
    player["buffDel"]                = &Player::buffDel;
    player["buffDelGroup"]           = &Player::buffDelGroup;
    player["buffDelMultiCount"]      = &Player::buffDelMultiCount;
    player["buffDelMultiGroupByID"]  = &Player::buffDelMultiGroupByID;
    player["buffSetLeftActiveCount"] = &Player::buffSetLeftActiveCount;
    player["buffSetNextActiveFrame"] = &Player::buffSetNextActiveFrame;
    player["cooldownClearTime"]      = &Player::cooldownClearTime;
    player["cooldownModify"]         = static_cast<void (Character::*)(int, int)>(&Character::cooldownModify);
    player["cooldownReset"]          = &Player::cooldownReset;
    player["skillGetLevel"]          = &Player::skillGetLevel;
    player["skillCast"]              = sol::overload(
        static_cast<void (Character::*)(int, int)>(&Character::skillCast),
        static_cast<void (Character::*)(int, int, int)>(&Character::skillCast),
        static_cast<void (Character::*)(int, int, int, int)>(&Character::skillCast)
    );
    player["skillCastXYZ"]     = &Player::skillCastXYZ;
    player["skillrecipeExist"] = &Player::skillrecipeExist;
    player["sceneGet"]         = &Player::sceneGet;
    player["timerSet"]         = sol::overload(
        static_cast<void (Character::*)(int, std::string, int)>(&Character::timerSet),
        static_cast<void (Character::*)(int, std::string, int, int)>(&Character::timerSet)
    );
    player["otherIsFormationLeader"]         = &Player::otherIsFormationLeader;
    player["otherIsInParty"]                 = &Player::otherIsInParty;
    player["otherGetKungfuMountID"]          = &Player::otherGetKungfuMountID;
    player["otherGetMapID"]                  = &Player::otherGetMapID;
    player["otherClearAllNpcThreat"]         = &Player::otherClearAllNpcThreat;
    player["otherCreatePublicShadow"]        = &Player::otherCreatePublicShadow;
    player["otherDestroyPublicShadow"]       = &Player::otherDestroyPublicShadow;
    player["otherDoAction"]                  = &Player::otherDoAction;
    player["otherPlayPublicShadowAnimation"] = &Player::otherPlayPublicShadowAnimation;
    player["dwID"]                           = &Player::dwID;
    player["nX"]                             = &Player::nX;
    player["nY"]                             = &Player::nY;
    player["nZ"]                             = &Player::nZ;
    player["nRoleType"]                      = &Player::nRoleType;
    player["nCurrentSunEnergy"]              = &Player::nCurrentSunEnergy;
    player["nCurrentMoonEnergy"]             = &Player::nCurrentMoonEnergy;
    player["nSunPowerValue"]                 = &Player::nSunPowerValue;
    player["nMoonPowerValue"]                = &Player::nMoonPowerValue;
    player["bSurplusAutoCast"]               = &Player::bSurplusAutoCast;
    player["bSurplusAutoReplenish"]          = &Player::bSurplusAutoReplenish;
    player["bFightState"]                    = &Player::bFightState;
    player["fMaxLife64"]                     = &Player::fMaxLife64;
    player["fCurrentLife64"]                 = &Player::fCurrentLife64;
    player["nLevel"]                         = sol::property(&Player::nLevel_getter, &Player::nLevel_setter);
    player["attrImportFromData"]             = &Player::attrImportFromData;
    player["attrImportFromJX3BOX"]           = &Player::attrImportFromJX3BOX;
    player["buffTimeLeftTick"]               = &Player::buffTimeLeftTick;
    player["itemAdd"]                        = static_cast<void (Character::*)(int, int)>(&Player::itemAdd);
    player["itemUse"]                        = static_cast<void (Character::*)(int, int)>(&Player::itemUse);
    player["cast"]                           = &Player::cast;
    player["skillActive"]                    = &Player::skillActive;
    player["skillDeactive"]                  = &Player::skillDeactive;
    player["skillLearn"]                     = &Player::skillLearn;
    player["skillrecipeAdd"]                 = &Player::skillrecipeAdd;
    player["skillrecipeRemove"]              = &Player::skillrecipeRemove;
    player["skilleventAdd"]                  = &Player::skilleventAdd;
    player["skilleventRemove"]               = &Player::skilleventRemove;
    player["dwKungfuID"]                     = &Player::kungfuID;
    player["publicCooldownID"]               = &Player::publicCooldownID;
    player["delayBase"]                      = &Player::delayBase;
    player["delayRand"]                      = &Player::delayRand;
    player["delayCustom"]                    = &Player::delayCustom;
    player["macroIdx"]                       = &Player::macroIdx;

    player["macroSkillCast"] = &Player::macroSkillCast;
    player["macroBuff"]      = &Player::macroBuff;
    player["macroNoBuff"]    = &Player::macroNoBuff;
    player["macroBufftime"]  = &Player::macroBufftime;
    player["macroTBuff"]     = &Player::macroTBuff;
    player["macroTNoBuff"]   = &Player::macroTNoBuff;
    player["macroTBufftime"] = &Player::macroTBufftime;
}

static void constructorAfter(CustomLua *self) {
    self->fightPrepareAdd = self->lua["Init"];
    int macroNum          = self->lua["MacroNum"].get<int>();
    for (int i = 0; i < macroNum; i++) {
        self->macroRuntime.push_back(self->lua["Macro" + std::to_string(i)]);
    }
}
